Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -254,7 +320,13 @@ extension SwiftLanguageServer {
       colorProvider: .bool(true),
       foldingRangeProvider: .bool(true),
       executeCommandProvider: ExecuteCommandOptions(
-        commands: builtinSwiftCommands)
+        commands: builtinSwiftCommands),
+      semanticTokensProvider: SemanticTokensOptions(
+        legend: SemanticTokensLegend(
```

Comment:
Good question, perhaps @DavidGoldman, who implemented the clangd forwarding IIUC, has some ideas on this?

Reply:
Is this related to why `SourceKitServer` responds without `semanticTokensProvider`? I'm talking about this response

https://github.com/apple/sourcekit-lsp/blob/1cb9e070ea8aeca381ba26a7847625287129ddad/Sources/SourceKitLSP/SourceKitServer.swift#L547-L553

Reply:
Dynamic registration lets us work around the fact that clangd has its own separate semantic token legends, distinct from the one we maintain for Swift - otherwise we'd have to unify them somehow (e.g. using clangd's or converting it to a single format). Since dynamic registration supports per-language registration (see [here](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#textDocumentRegistrationOptions)), it gives us semantic highlighting for both ObjC/Swift as long as the editor supports it.

In the short term, for editors that don't support dynamic registration, we could maybe do something like I've done [here](https://github.com/apple/sourcekit-lsp/pull/429/files#diff-74f7ae001022599e910e900d4fb40255212cfe8887ad1322918aa2657b733e46R531) to at least give us support for Swift semantic highlighting.

Reply:
> Does this answer your question?

thank you both! it does answer my question.

Side note: It does feel like sourcekit-lsp is a bit of stretch for the LSP specification while trying to bridge `clangd`. Maybe we're close to the point where sourcekit-lsp should have launch argument to enable/disable clangd - this would make it more predictable.